pennylane/gradients/gradient_transform.py

```diff
@@ -108,6 +108,26 @@ def __init__(
         self.hybrid = hybrid
         super().__init__(transform_fn, expand_fn=expand_fn, differentiable=differentiable)
 
+    @staticmethod
+    def _jacobian_trainable_args(args, interface):
+        """Return the indices of QNode arguments for which a Jacobian was
+        computed by `qml.gradients.classical_jacobian` with argnum=None.
+        """
+        trainable_args = []
+
+        if interface == "autograd":
+            for idx, arg in enumerate(args):
+                if getattr(arg, "requires_grad", True):
+                    trainable_args.append(idx)
+
+        elif interface == "jax":
+            trainable_args = [0]
+
+        else:
+            trainable_args = range(len(args))

```

Comment:
To answer your question, that is defined by the behaviour of `qml.transforms.classical_jacobian` (and the respective interfaces) which I mimic here. The gradient transform exclusively [calls](https://github.com/PennyLaneAI/pennylane/blob/e4dc1d2e8e31544bcc7abc4c97c79c0cbb98816c/pennylane/gradients/gradient_transform.py#L117) `classical_jacobian` with `argnum=None` and I inferred its behaviour from the source code and the table in its [docs page](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.transforms.classical_jacobian.html). 

Something important I forgot to mention is that this patch essentially only applies to autograd (also in regards to your last two comments). I wrote this function to help me determine which behaviour was exhibited by `qml.jacobian` in regards to stacking multiple "single-argument" Jacobians, which I do by examining the shape of the classical Jacobian `cjac` (when it is not a tuple of multiple Jacobians). It's not an ideal solution but one that avoids modifying the signature/public interface of `qml.jacobian`.

In regards to the other interfaces, looking at the [docs table](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.transforms.classical_jacobian.html) for `classical_jacobian`, TensorFlow and Torch will _always_ return a tuple of Jacobians for `cjac`, so this function is never invoked there. I originally had this function as follows, just to provide sensible values for each interface even if they are unused:
```py
        trainable_args = []

        if interface == "autograd":
            for idx, arg in enumerate(args):
                if getattr(arg, "requires_grad", True):
                    trainable_args.append(idx)

        elif interface == "torch":
            trainable_args = range(len(args))

        elif interface == "jax":
            trainable_args = [0]

        elif interface == "tf":
            trainable_args = range(len(args))

        return trainable_args
```
To satisfy codecov I combined torch and tf into the else branch and added a "dummy" test for it (testing this function on its own doesn't provide much value I don't think).
Note the jax branch does get invoked and it's always the "single argument" case, because with `argnum=None` it always returns the Jacobian with respect to the first argument.

Reply:
> In regards to the other interfaces, looking at the docs table for classical_jacobian, TensorFlow and Torch will always return a tuple of Jacobians for cjac, so this function is never invoked with them.

I'd be tempted to remove the `else` branch then and mention this in the docstring of the method.

Indeed, looking at `classical_jacobian`, I'm somewhat surprised. With Torch, even those arguments that have `requires_grad=False` set get their classical jacobian computed. :thinking: